src/validators/dataclass.rs

```diff
@@ -516,14 +529,21 @@ impl Validator for DataclassValidator {
         if self.frozen {
             return Err(ValError::new(ErrorType::FrozenInstance, field_value));
         }
-        let dict_py_str = intern!(py, "__dict__");
-        let dict: &PyDict = obj.getattr(dict_py_str)?.downcast()?;
 
-        let new_dict = dict.copy()?;
+        let new_dict = if let Some(ref slots) = self.slots {
+            let slots_dict = PyDict::new(py);
+            for slot in slots {
+                let slot = slot.as_ref(py);
+                slots_dict.set_item(slot, obj.getattr(slot)?)?;
+            }
+            slots_dict
+        } else {
+            let dunder_dict: &PyDict = obj.getattr(intern!(py, "__dict__"))?.downcast()?;
+            dunder_dict.copy()?
+        };
+
         new_dict.set_item(field_name, field_value)?;
```

Comment:
`new_dict` here seems wrong in the case that there are both slots and non-slots fields in the dataclass.